cmd2/AutoCompleter.py

```diff
@@ -0,0 +1,796 @@
+
+        def consume_flag_argument() -> None:
+            """Consuming token as a flag argument"""
+            # we're consuming flag arguments
+            # if this is not empty and is not another potential flag, count towards flag arguments
+            if len(token) > 0 and not token[0] in self._parser.prefix_chars and flag_action is not None:
+                flag_arg.count += 1
+
+                # does this complete a option item for the flag
+                arg_choices = self._resolve_choices_for_arg(flag_action)
+                # if the current token matches the current position's autocomplete argument list,
+                # track that we've used it already.  Unless this is the current token, then keep it.
+                if not is_last_token and token in arg_choices:
+                    consumed_arg_values.setdefault(flag_action.dest, [])
+                    consumed_arg_values[flag_action.dest].append(token)
+
+        def consume_positional_argument() -> None:
+            """Consuming token as positional argument"""
+            pos_arg.count += 1
+
+            # does this complete a option item for the flag
+            arg_choices = self._resolve_choices_for_arg(pos_action)
+            # if the current token matches the current position's autocomplete argument list,
+            # track that we've used it already.  Unless this is the current token, then keep it.
+            if not is_last_token and token in arg_choices:
+                consumed_arg_values.setdefault(pos_action.dest, [])
+                consumed_arg_values[pos_action.dest].append(token)
+
+        is_last_token = False
+        for idx, token in enumerate(tokens):
+            is_last_token = idx >= len(tokens) - 1
+            # Only start at the start token index
+            if idx >= self._token_start_index:
+                current_is_positional = False
+                # Are we consuming flag arguments?
+                if not flag_arg.needed:
+                    # we're not consuming flag arguments, is the current argument a potential flag?
+                    if len(token) > 0 and token[0] in self._parser.prefix_chars and\
+                            (is_last_token or (not is_last_token and token != '-')):
+                        # reset some tracking values
+                        flag_arg.reset()
+                        # don't reset positional tracking because flags can be interspersed anywhere between positionals
+                        flag_action = None
+
+                        # does the token fully match a known flag?
+                        if token in self._flag_to_action.keys():
+                            flag_action = self._flag_to_action[token]
+                        elif hasattr(self._parser, 'allow_abbrev') and self._parser.allow_abbrev:
+                            candidates_flags = [flag for flag in self._flag_to_action.keys() if flag.startswith(token)]
```

Comment:
In Python, iterating through a dictionary just iterates through the keys, so no need to call the **.keys()** method, can just iterate through the dictionary directly.